keps/sig-scheduling/5229-asynchronous-api-calls-during-scheduling/README.md

```diff
@@ -0,0 +1,998 @@
+- Status update (patch): Apply newest API call
+- Binding: Ignore status update API calls
+- Delete (in preemption): Ignore status update as well as binding API calls
+
+```go
+type APICallType string
+
+const (
+  StatusUpdate APICallType = "status_update"
+  Binding      APICallType = "binding"
+  Delete       APICallType = "delete"
+)
+
+type PodAPICall struct {
+	podID    types.UID
+	callType APICallType
+	fn       func()
+}
+
+type APIQueue struct {
+  ...
+}
+
+func (aq *APIQueue) Add(podAPICall PodAPICall) {
+  // If API call for specific podID is already enqueued,
+  // check the callType and skip or replace the call depending on precedence.
+  ...
+}
+
+func (aq *APIQueue) Run() {
+  // Dispatch limited number of goroutines if queue is non empty.
+  ...
+}
+```
+
+Pros:
+- Allows for advanced goroutine dispatching logic.
+- Can potentially delay, skip, or merge API calls based on type (e.g., skip `nominatedNodeName` if binding is pending).
+- All collisions could be resolved at the queue level, not relying on higher-level mechanisms (like 1.1 or 1.2).
+- Allows for (1.3) where all scenarios can be supported without additional structures.
+- Provides a framework that can be extended in the future.
+
+Cons:
+- Requires complex logic to handle potential conflicts between different update types for the same pod.
+- Needs a clear strategy for how to update the in-memory pod object during scheduling.
+
+
+#### 2.3: Send API calls through a kube-scheduler's cache
+
+A third approach could be to have a consistent pod state in the kube-scheduler itself first and then change it through the API.
+This means that all API calls whould have to go through the kube-scheduler's cache, change the pod there, and after that, execute.
+However, pod updates might come from outside the kube-scheduler, e.g., a user changes the spec or something changes the status (if it is even possible).
+This extended cache would have to merge the internal state of the pod with the external state,
+including the pod update made by the kube-scheduler that will come as an event as well.
+Now, the pod object stored in the cache is based only on events that come to the kube-scheduler.
+
+Another thing to think of is that the cache stores only the bound pods. The rest of the pods is stored in the scheduling queue,
+so once again, API calls might need to go through the scheduling queue itself.
+
+Pros:
+- Aims for a consistent internal state of the pod within the kube-scheduler before calling the API, possibly simplifying conflict resolution.
+
+Cons:
+- Requires the cache to handle and merge updates coming from both the kube-scheduler's internal actions and external API events.
+- The cache currently only stores bound pods, requiring integration with the scheduling queue for pending pods.
+- Complex logic is needed to handle external updates arriving while an internal update is pending or in progress.
```

Comment:
Regarding the first and the third cons: They're mentioned as cons here. However, this is a general problem, not a problem specific to this option since no option gracefully solves: how to handle the conflicts between changes the scheduler wants to make (but awaiting the scheduler to actually make an API call) vs the changes coming from outside. 

When such a conflict happens, which updates should be prioritized depends on which fields have to be updated on what purpose. That's a problem that I don't see an answer yet.